### ISSUE TITLE
Make Player.SetRunSpeedMultThisFrame take values over 1.49

### ISSUE THREAD

**Issue by justalemon** [scripthookvdotnet/scripthookvdotnet#1317]
Native documentation says

> Multiplier goes up to 1.49 any value above will be completely overruled by the game and the multiplier will not take effect, this can be edited in memory however.

I was wondering if it would be possible for the Player.SetRunSpeedMultThisFrame function to take values over 1.49 by using memory patters or offsets instead of SET_RUN_SPRINT_MULTIPLIER_FOR_PLAYER.

**Comment by kagikn** [comment]
You may want to see how the player moves when you set the run mult to a higher value such as 2.0 before we would directly modify the value by difect memory access. We'll have to find a memory pattern if someone wants to get the current value though. SET_RUN_SPRINT_MULTIPLIER_FOR_PLAYER basically does nothing if the passed value is not in the range of [1.0, 1.5), so no, you can't set the run mult to 1.5 or above (that native doc says slightly incorrect info about the upper bound). But it looks like Rockstar limits the range for good reasons...

Anyway, try messing with the run multiplier with this snippet, and I'd love to receive your feedback. If you want to get the run mult on the fly, we'd need to find a memory pattern though.
```cs
void SetPlayerRunMultNoRangeLimit(Player player, float mult)
{
    IntPtr address = player.MemoryAddress;

    const offset = 0xD40 // for b2944, use 0xCF0 for b2699

    // bruh, BCL doesn't have Marshal.WriteSingle even in .NET 7
    byte[] floatBytes = BitConverter.GetBytes(mult);
    Marshal.Copy(floatBytes, 0, player.MemoryAddress + offset, floatBytes.Length);

    // if you are willing to use unsafe block, you can use the line below instead of the 2 lines above.
    // I'm assuming you don't like to use them considering how LemonUI is coded though...
    /* *(float*)(Game.Player.MemoryAddress + offset) = mult; */
}
```

If you are using v3.6.0, you can find a `CPlayerInfo` offset from the lines referenced below (the offset variable is named `PedPlayerInfoOffset`). Use the offset like `Marshal.ReadIntPtr(Game.Player.MemoryAddress + PedPlayerInfoOffset)`.
https://github.com/scripthookvdotnet/scripthookvdotnet/blob/c9e31faeb3d0a8f5a260a18eb92645df76e2241d/source/core/NativeMemory.cs#L574C2-L579

**Comment by justalemon** [comment]
> I'm assuming you don't like to use them considering how LemonUI is coded though...

That was mainly to make sure that I wouldn't accidentally commit unsafe code to platforms that don't support it like RageMP and FiveM.

**Comment by kagikn** [comment]
Yeah, that makes sense, avoiding any unsafe code when you're building libraries that support platforms where unsafe code isn't supported.
I'm looking forward to your feedback about how the player moves when the run multiplier is 1.5 or more too. Alternatively, you can explain how the getter would be useful as it requires to fetch the offset. Concrete examples would convince me or others more I guess? They would form a more concrete consensus about if finding a memory pattern for the offset that SET_RUN_SPRINT_MULTIPLIER_FOR_PLAYER uses would worth the effort even if the nearby assembly code changes and we need to build another pattern again. Test how the player moves at higher values like 2.0 and leave your feedback, or explain how the getter would be useful 😛
